### PR TITLE
test: a test to check descendant limits

### PR DESCRIPTION
**Pull request by ritickgoenka**
This PR adds a new functional test to test the new descendant limits for packages that were proposed in #21800. 
 ```
+----------------------+
|                      |
|         M1           |
|        ^  ^          |
|       M2   ^         |
|      .      ^        |
|     .        ^       |
|    .          ^      |
|   .            ^     |
|  M24            ^    |
|                  ^   |
|                  P1  |
|                  ^   |
|                  P2  |
|                      |
+----------------------+
```                                                                     
                                     
This test is for checking a transaction to fail its descendant count limits because of a combination of mempool descendants, package direct descendants, and package indirect descendants.

In this test,  P1 has M1 as a mempool ancestor, P2 has no in-mempool ancestors, but when combined P2 has M1 as an ancestor and M1 exceeds descendant_limits (23 in-mempool descendants + 2 in-package descendants, a total of 26 including itself)

### CODE REVIEW

**Comment by tryphe** [comment]
Concept ACK

**Review by glozow — COMMENTED** [review]
Concept ACK to more testing.

The commit message for c18188e96204f9264e2ed64ab5e53f3abdd137e5 seems to contain some squash comments? Would be better to describe exactly what edge case you're testing / coverage you're adding.

File: test/functional/rpc_packages.py [hunk 1]
```diff
@@ -304,6 +305,109 @@ def test_descendant_limits(self):
         assert all([res["allowed"] for res in node.testmempoolaccept(rawtxs=package_hex)])
         node.generate(1)
 
+    def test_descendant_limits_bushy(self):
```
[the remark this answers is not in the document]
Comment (ritickgoenka) [reply test/functional/rpc_packages.py]: Done

File: test/functional/rpc_packages.py [hunk 2]
```diff
@@ -304,6 +305,109 @@ def test_descendant_limits(self):
...
+                ^       ^            M23a
+                 ^     ^
+                  ^   P1
+                   ^ ^  ^
+                    P2   P3
+
+        P1 has M1a as a mempool ancestor, P2 has only M1b as in-mempool ancestor whereas P3 has
+        no in-mempool ancestors, but when combined both P2 and P3 have M1a as an ancestor and
+        M1a exceeds descendant_limits (23 in-mempool descendants + 3 in-package descendants)
+        """
```
[the remark this answers is not in the document]
Comment (ritickgoenka) [reply test/functional/rpc_packages.py]: To test a transaction having a direct connection to mempool doesn't affect it being counted as an indirect descendant of some transaction in the mempool. If this doesn't make much sense I can remove M1b.

**Comment by glozow** [comment]
Will review after rebase - note that the mempool tests have been moved to mempool_package_limits.py

File: test/functional/mempool_package_limits.py [hunk 3]
```diff
@@ -177,6 +178,107 @@ def test_desc_count_limits(self):
...
+        """
+
+        node = self.nodes[0]
+        package_hex = []
+        parents_tx = []
+        values = []
+        scripts = []
+        # M1a
+        first_coin_a = self.coins.pop()
+        parent_value = (first_coin_a["amount"] - Decimal("0.0002")) / 2 # Deduct reasonable fee and make 2 outputs
```
Comment (glozow) [inline test/functional/mempool_package_limits.py:206]: I actually think a fee of just `Decimal("0.0001")` would be sufficient here. In fact, I think it'd be best to import `DEFAULT_FEE` from test_framework/wallet.py and use that throughout this test.

File: test/functional/mempool_package_limits.py [hunk 4]
```diff
@@ -177,6 +178,107 @@ def test_desc_count_limits(self):
...
+                             ^  ^
+                            ^    M2a
+            M1b            ^       .
+              ^           ^         .
+               ^         ^           ^
+                ^       ^            M23a
+                 ^     ^
+                  ^   P1
+                   ^ ^  ^
+                    P2   P3
```
Comment (glozow) [inline test/functional/mempool_package_limits.py:193]: nit: this test would be simplified if you made M2a .... M23a be a chain spending vout[0] of M1a. It'd save maybe 5-10 lines.

File: test/functional/mempool_package_limits.py [hunk 5]
```diff
@@ -177,6 +178,107 @@ def test_desc_count_limits(self):
         node.generate(1)
         assert all([res["allowed"] for res in node.testmempoolaccept(rawtxs=package_hex)])
 
+    def test_desc_count_limits_2(self):
+        """Create A bushy Package with 24 transaction in mempool and 3 transaction in package
```
Comment (glozow) [inline test/functional/mempool_package_limits.py:182]: Again, not bushy

**Review by glozow — COMMENTED** [review]
Thanks for rebasing, the test is well-written and does what the comment says it does.

Suggestion: label your commit message with [test] and add a short description about why this test is an improvement.

File: test/functional/rpc_packages.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (glozow) [reply test/functional/rpc_packages.py]: ^ This description or a short summary of it would be good in place of the "bushy" comment

**Comment by glozow** [comment]
Another suggestion (probably for a different PR): This doesn't matter _that_ much, but we have quite a few of the same lines repeated in this test. Given how many OP_TRUE spends there are, I think it'd be good to either add a OP_TRUE variation of `make_chain` to wallet.py or, better yet, make all those functions methods of `MiniWallet`.

File: test/functional/rpc_packages.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (glozow) [reply test/functional/rpc_packages.py]: Right now the test doesn't utilize M1b at all - it should test that the package [P1, P2] passes but [P1, P2, P3] fails.

File: test/functional/mempool_package_limits.py [hunk 6]
```diff
@@ -177,6 +179,102 @@ def test_desc_count_limits(self):
...
+        values.append(value)
+        scripts.append(spk)
+
+        # P2
+        child_hex = create_child_with_parents(node, self.address, self.privkeys, parents_tx, values, scripts)
+        package_hex.append(child_hex)
+
+        assert_equal(24, node.getmempoolinfo()["size"])
+        assert_equal(3, len(package_hex))
+        testres = node.testmempoolaccept(rawtxs=package_hex)
```
Comment (glozow) [inline test/functional/mempool_package_limits.py:270]: Did you consider the suggestion about asserting success for [P1, P3] but failure for [P1, P2, P3] ?

File: test/functional/mempool_package_limits.py [hunk 7]
```diff
@@ -177,6 +179,102 @@ def test_desc_count_limits(self):
...
+                 .          ^       ^
+                M23a         ^     ^
+                              ^   ^
+                               ^ ^
+                               P1
+                              ^  ^
+                             P2   P3
+        P1 has M1a as a mempool ancestor, P2 has only M1b as in-mempool ancestor whereas P3 has
+        no in-mempool ancestors, but when combined both P2 and P3 have M1a as an ancestor and
+        M1a exceeds descendant_limits (23 in-mempool descendants + 3 in-package descendants)
```
Comment (glozow) [inline test/functional/mempool_package_limits.py:198]: This comment/diagram is now incorrect - M1b is the parent of P2, not P1

File: test/functional/mempool_package_limits.py (hunk 6 above)
Comment (ritickgoenka) [reply test/functional/mempool_package_limits.py:252]: I have removed M1b as it was creating more confusion and as jnewbery pointed out the test would still fail regardless of M1b being present there and it was not adding any edgecase that it not already present.

File: test/functional/mempool_package_limits.py [hunk 8]
```diff
@@ -177,6 +179,84 @@ def test_desc_count_limits(self):
...
+                  .        ^
+                 .          ^
+                M23a         ^
+                              ^
+                              P1
+                             ^  ^
+                            P2   P3
+        P1 has M1a as a mempool ancestor, P2 and P3 have no in-mempool ancestors, but when
+        combined both P2 and P3 have M1a as an ancestor and M1a exceeds descendant_limits
+        (23 in-mempool descendants + 3 in-package descendants)
```
Comment (glozow) [inline test/functional/mempool_package_limits.py:197]:
```suggestion
        (22 in-mempool descendants + 3 in-package descendants, a total of 26 including itself).
```

**Review by glozow — COMMENTED** [review]
utACK 5a7050c63c1a574abf336a47b18e154554a4020a

**Comment by glozow** [comment]
PR description needs to be updated

**Comment by jnewbery** [comment]
I still don't understand what logic this is testing that isn't already covered by the `test_desc_count_limits()` subtest.

**Comment by ritickgoenka** [comment]
> I still don't understand what logic this is testing that isn't already covered by the `test_desc_count_limits()` subtest.

It is testing if the second generation in Package transactions are considered while checking Descendant count limits. It is analogous to the 'test_anc_count_limits_2()' subtest for Ancestor count limits.

**Comment by jnewbery** [comment]
> It is testing if the second generation in Package transactions are considered while checking Descendant count limits.

What do you think about simplifying the test to:

```
      ----------------
             M1
              |
             M2
              .
  Mempool     .
              .
             M23
              ^
      ----------------
              ^
             P1
  Package     |
             P2
      ----------------
```

That seems to capture the scenario that you're trying to test in the simplest form.

I also recommend that you remove the image from the PR description, and replace it with the ascii drawing. The PR description is included in the merge commit, so it's best if it's just plain text.

**Comment by jnewbery** [comment]
@glozow has pointed out that this scenario is for a transaction to fail its descendant limit count due to a combination of mempool descendants, package direct descendants, and package indirect descendants, which seems like a reasonable thing to test.

I think the transaction P3 is unnecessary, since the package would fail with just P2. It's good for a test to fail at the edge case (i.e. the package would be accepted if you removed just one transaction).

I'd also recommend you remove the image from the PR description and replace it with an ascii drawing. The PR description is included in the merge commit, so it's best if it's plain text without attachments.

File: test/functional/mempool_package_limits.py [hunk 9]
```diff
@@ -177,6 +179,73 @@ def test_desc_count_limits(self):
...
+        tx_child_p1_spk = tx_child_p1.vout[0].scriptPubKey.hex()
+
+        # P2
+        (_, tx_child_p2_hex, _, _) = make_chain(node, self.address, self.privkeys, txid_child_p1, value_p1, 0, tx_child_p1_spk)
+        package_hex.append(tx_child_p2_hex)
+
+        assert_equal(24, node.getmempoolinfo()["size"])
+        assert_equal(2, len(package_hex))
+        testres = node.testmempoolaccept(rawtxs=package_hex)
+        for txres in testres:
```
Comment (ryanofsky) [inline test/functional/mempool_package_limits.py:242]: In commit "[test] checks descendants limtis for second generation Package descendants" (e0f92f30954438540e6a676af77d17fbb963ca5d)

May be paranoid, but could be good to check `assert_equal(len(testres), len(package_hex))`

**Review by ryanofsky — COMMENTED** [review]
Code review ACK e0f92f30954438540e6a676af77d17fbb963ca5d. Seems like a useful and straightforward test case, assuming it doesn't duplicate existing coverage.

**Review by ryanofsky — APPROVED** [review]
Code review ACK fa7db1cbf7e8400b625fccd5757f8e1b200796bd. Only were suggested changes since last review: simplifying test and dropping P3 transaction as John suggested, and adding assert_equal I suggested

**Review by glozow — COMMENTED** [review]
ACK fa7db1cbf7e8400b625fccd5757f8e1b200796bd

File: test/functional/mempool_package_limits.py [hunk 10]
```diff
@@ -177,6 +179,74 @@ def test_desc_count_limits(self):
...
+        parent_tx = tx_from_hex(parent_signed["hex"])
+        parent_txid = parent_tx.rehash()
+        node.sendrawtransaction(parent_signed["hex"])
+
+        # Chain M2...M24
+        spk = parent_tx.vout[0].scriptPubKey.hex()
+        value = parent_value
+        txid = parent_txid
+        for i in range(23): # M2...M24
+            (tx, txhex, value, spk) = make_chain(node, self.address, self.privkeys, txid, value, 0, spk)
```
Comment (jnewbery) [inline test/functional/mempool_package_limits.py:220]:
```suggestion
            tx, txhex, value, spk = make_chain(node, self.address, self.privkeys, txid, value, 0, spk)
```

File: test/functional/mempool_package_limits.py [hunk 11]
```diff
@@ -177,6 +179,74 @@ def test_desc_count_limits(self):
...
+        descendants + 2 in-package descendants, a total of 26 including itself).
+        """
+
+        node = self.nodes[0]
+        package_hex = []
+        # M1
+        first_coin_a = self.coins.pop()
+        parent_value = (first_coin_a["amount"] - DEFAULT_FEE) / 2 # Deduct reasonable fee and make 2 outputs
+        inputs = [{"txid": first_coin_a["txid"], "vout": 0}]
+        outputs = [{self.address : parent_value}, {ADDRESS_BCRT1_P2WSH_OP_TRUE : parent_value}]
```
Comment (jnewbery) [inline test/functional/mempool_package_limits.py:206]: Consider using a linter like [flake8](https://flake8.pycqa.org/en/latest/) or a formatter like [black](https://github.com/psf/black) to catch/fix PEP8 violations:

```suggestion
        outputs = [{self.address: parent_value}, {ADDRESS_BCRT1_P2WSH_OP_TRUE: parent_value}]
```

**Review by jnewbery — COMMENTED** [review]
ACK fa7db1cbf7

I've left a couple of comments inline, but they can be picked up in #22901.

If you have to touch this again, I'd suggest changing the commit log:

```diff
-[test] checks descendants limtis for second generation Package descendants
+[test] check descendant limits for second generation package descendants
```
